zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+    let mut stmt = conn.prepare(
+        "SELECT address, address_index FROM ephemeral_addresses WHERE account_id = :account ORDER BY address_index",
+    )?;
+    let mut rows = stmt.query(named_params! { ":account": account_id.0 })?;
+
+    let mut result = HashMap::new();
+    let mut first_unused_index: Option<i32> = Some(0);
+
+    while let Some(row) = rows.next()? {
+        let addr_str: String = row.get(0)?;
+        let raw_index: u32 = row.get(1)?;
+        first_unused_index = i32::try_from(raw_index)
+            .map_err(|e| SqliteClientError::CorruptedData(e.to_string()))?
+            .checked_add(1);
+        let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+        let address = TransparentAddress::decode(params, &addr_str)?;
+        result.insert(address, Some(metadata(address_index)));
+    }
+
+    if for_detection {
+        if let Some(first) = first_unused_index {
+            let ephemeral_ivk = get_ephemeral_ivk(conn, params, account_id)?;
+
+            for raw_index in range_after(first, GAP_LIMIT) {
+                let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+                let address = ephemeral_ivk.derive_ephemeral_address(address_index)?;
+                result.insert(address, Some(metadata(address_index)));
+            }
+        }
+    }
+    Ok(result)
+}
+
+/// Returns a vector with the next `n` previously unreserved ephemeral addresses for
+/// the given account.
+///
+/// Precondition: `n < 0x80000000`
+///
+/// # Errors
+///
+/// * `SqliteClientError::AccountUnknown`, if there is no account with the given id.
+/// * `SqliteClientError::UnknownZip32Derivation`, if the account is imported and
+///   it is not possible to derive new addresses for it.
+/// * `SqliteClientError::ReachedGapLimit`, if it is not possible to reserve `n` addresses
+///   within the gap limit after the last address in this account that is known to have an
+///   output in a mined transaction.
+/// * `SqliteClientError::AddressGeneration(AddressGenerationError::DiversifierSpaceExhausted)`,
+///   if the limit on transparent address indices has been reached.
+pub(crate) fn reserve_next_n_ephemeral_addresses<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    account_id: AccountId,
+    n: u32,
+) -> Result<Vec<(TransparentAddress, TransparentAddressMetadata)>, SqliteClientError> {
+    if n == 0 {
+        return Ok(vec![]);
+    }
+    assert!(n > 0);
+    let n = i32::try_from(n).expect("precondition violated");
+
+    let ephemeral_ivk = get_ephemeral_ivk(wdb.conn.0, &wdb.params, account_id)?;
+    let last_reserved_index = last_reserved_index(wdb.conn.0, account_id)?;
+    let last_safe_index = last_safe_index(wdb.conn.0, account_id)?;
+    let allocation = range_after(last_reserved_index, n);
+
+    if allocation.clone().count() < n.try_into().unwrap() {
```

Comment:
It seems like a very weird hole that `ExactSizeIterator` is only implemented for `RangeInclusive<{u,i}{8,16}>`, so we can't use `allocation.len()` here.

Reply:
It's because the length might be exactly $2^{32}$ on a system with 32-bit `usize`: https://github.com/rust-lang/rust/issues/36386

Reply:
I ended up changing all index variables to unsigned and was able to use `.len()` here.